Added Chlorine `Cl1s` atomType

alongd opened the pull request:
This PR allows Chlorine to be reactive in RMG. Only valance 1 is considered, and no hypervalance atomTypes were defined.
A respective atomType test was added, and the documentation was changed accordingly.
I also made a small change to the isomorphismTest which otherwise insists that Cl should be charged.

This PR will soon be accompanied by a sister -db PR with relevant kinetics and thermo data by @zjburas. They should be merged together, but this branch can principally be reviewed.

mliu49 commented:
@rwest mentioned that he was working on chlorine in #717. Are there any updates on the status of that?

alongd commented:
@rwest , regarding @mliu49's comment above, are there any chlorine additions you'd like to add to either this PR or https://github.com/ReactionMechanismGenerator/RMG-database/pull/257? Or shall we continue with these as they are?

rwest commented:
Sorry, I've not been keeping on top of this or https://github.com/ReactionMechanismGenerator/RMG-database/pull/257. Here's a brief recap. 

Our work was done by @faribas  and can be found on her `halogens` branches. If she still receives github notifications, she can correct what I get wrong...

* RMG-Py: https://github.com/ReactionMechanismGenerator/RMG-Py/compare/master...faribas:halogens
* Database: https://github.com/ReactionMechanismGenerator/RMG-database/compare/master...faribas:halogens

Because this forked three years ago (March 2015) it's probably hard to merge them directly, but there may be value in copying across what she did. 

**Thermochemistry:**
as well as adding some thermo libraries and Benson's group additivity values for chlorinated hydrocarbons, she found she had to calculate her own hydrogen bond increment (HBI) values in order to estimate radicals containing chlorines. Next-nearest-neighbour effects were required to get decent estimates. She derived these from her own quantum calculations.

**Reactions:**
She allowed Cl to react as the radical for the two families _H abstraction_ and _radical recombination_, and created new reaction families for _H-Cl or Cl-Cl insertion into a multiple bond_ and _Cl abstraction_.  Training rates were from the literature.

More details are in chapter 3 of her PhD thesis at http://hdl.handle.net/2047/D20213055 

Also worth mentioning is that I have just received a grant to work on adding halogens to RMG (https://www.nsf.gov/awardsearch/showAward?AWD_ID=1751720), so would love to start a conversation about the best way to do this! It will be @skrsna working on that project, starting soon. Perhaps we can all meet.

alongd commented:
Congrats on the grant! I love that this includes training undergrads in this kind of problem solving.

Other stakeholders here are @zjburas (currently in CA), and @sarahkha from the GreenGroup implementing iodine.

I'll look at @faribas's -Py repo, @zjburas, could you take a look at the -db repo?

zjburas commented:
Sure @alongd , I can take a look at the -db repo. 

Congratulations @rwest on the grant. My immediate interest in halogen chemistry is pretty narrow: I just want to use RMG as a tool to account for chlorine side-chemistry during chlorine initiated oxidation experiments from 300-1000 K. For that purpose, what @faribas already did is probably sufficient. Hopefully we can talk sometime.

amarkpayne commented on documentation/source/reference/molecule/atomtype.rst:
```diff
@@ -116,3 +116,8 @@ Atom type       Description
 ``S6tt``        sulfur atom with no lone pairs (valance 6) and two triple bonds
 ``S6tdc``       charged sulfur atom with no lone pairs (valance 6), one to two triple bonds, up to two double bonds, and up to four single bonds
 =============== ==============================================================================================================================================================
```
You need to delete line 118 for the table to format properly. Otherwise, it tries to end the table early

amarkpayne commented on documentation/source/users/rmg/database/introduction.rst:
```diff
@@ -189,6 +189,8 @@ table below shows all atoms types in RMG.
 +----------+-------------------+------------------------------------------------------------------------------------------------------------------+
 |S6tdc     |Sulfur             |No lone pairs, one to two triple bonds, up to two double bonds, up to four single bonds, charged -1/-1            |
 +----------+-------------------+------------------------------------------------------------------------------------------------------------------+
+|Cl1s      |Chlorine           |Three lone pairs, zero to one single bonds                                                                        |
```
Should you add an entry for Cl as well?

amarkpayne commented on rmgpy/cantherm/statmech.py:
```diff
@@ -705,10 +705,10 @@ def applyEnergyCorrections(E0, modelChemistry, atoms, bonds):
-            'C#C': -0.64,  'C#N': -0.89,  'C-S': 0.43,   'O=S': -0.78, 'S-H': 0.0,   'C-N': -0.13, # Table IX: Petersson GA (1998) J. of Chemical Physics, DOI: 10.1063/1.477794 
-            'N-H': -0.42,  'N=O': 1.11,   'N-N': -1.87,  'N=N': -1.58, 'N-O': 0.35, #Table 2: Ashcraft R (2007) J. Phys. Chem. B; DOI: 10.1021/jp073539t
-            'N#N': -2.0,   'O=O': -0.2,   'H-H': 1.1, # Unknown source
+            'C-C': -0.495,'C-H': -0.045,'C=C': -0.825,'C-O': 0.378,'C=O': 0.743,'O-H': -0.423,  #Table2: Paraskevas, PD (2013). Chemistry-A European J., DOI: 10.1002/chem.201301381
```
It looks like your editor changed some of the spacing here

alongd replied:
That was actually me, trying to make the lines a bit shorter (now with the 'C-Cl', 'C-F' additions, the second line was too long). It still functions good and it's readable, let me know if you think otherwise.

amarkpayne replied:
Gotcha, this is fine by me then, just making sure

amarkpayne commented on rmgpy/molecule/group.py:
```diff
@@ -1117,18 +1117,25 @@ def updateFingerprint(self):
+        self.siliconCount = 0
         self.radicalCount  = 0
         for atom in self.vertices:
             if len(atom.atomType) == 1:
```
This is outside the scope of your commit, but what should happen if len(atom.atomType) != 1?

alongd replied:
My understanding is that since it is a Group and not a Molecule, an "atom" can be defined as `1 [C,N,O] ux px cx {2,S}`, in which case we can't fingerprint it and we correctly skip this condition. I think this also means we're missing out on cases like `1 [Cs,Cd] ux px cx {2,S}` where we could fingerprint, but don't (for that atom in the Group). However, this should be working smoothly for Molecules.

amarkpayne replied:
That makes sense. I wonder if there is a way to catch cases like `1 [Cs,Cd] ux px cx {2,S}` and if that would be useful. If so this is something for another PR though.

amarkpayne commented on rmgpy/molecule/group.py:
```diff
@@ -1137,17 +1144,24 @@ def updateFingerprint(self):
+                isChlorine = atomType.equivalent(chlorine)
+                isSilicon  = atomType.equivalent(silicon)
+                sum_is_atom = isCarbon + isNitrogen + isOxygen + isSulfur + isChlorine + isSilicon
+                if sum_is_atom == 1:
```
What happens if sum_is_atom != 1? I can't imagine this happening, but maybe I am missing a subtlety here. I would think that this would either always evaluate to `True` (and thus it is not needed) or we need some sort of error handling if it does occur.

alongd replied:
I think this is related to the previous comment: an "atom" in a Group could be defined as `1 [C,N,O] ux px cx {2,S}`. In this case, atomType = [C,N,O], and `.equivalent` will return `True` for more than one isAtom. In this case, `sum_is_atom` will be 3.
I agree that it seems unnecessary (now that I rethink it) since it's already under the `if len(atom.atomType) == 1:` condition discussed above. We could either keep it to keep the same functionality as before in case I'm also missing a subtlety, or remove. I'll check.

amarkpayne commented on rmgpy/molecule/molecule.pxd:
```diff
@@ -61,7 +61,15 @@ cdef class Atom(Vertex):
 
+    cpdef bint isFluorine(self)
+
+    cpdef bint isSilicon(self)
```
Make a note in the commit that `isSilicon` was added to cython file as well, good catch!

amarkpayne commented on rmgpy/molecule/molecule.py:
```diff
@@ -307,6 +307,13 @@ def isOxygen(self):
 
+    def isFluorine(self):
+        """
+        Return ``True`` if the atom represents an oxygen atom or ``False`` if
```
Change doc string element name to "fluorine"

amarkpayne commented on rmgpy/molecule/molecule.py:
```diff
@@ -321,6 +328,20 @@ def isSulfur(self):
 
+    def isChlorine(self):
+        """
+        Return ``True`` if the atom represents an sulfur atom or ``False`` if
```
Same here, but for "chlorine"

amarkpayne commented on rmgpy/molecule/molecule.py:
```diff
@@ -321,6 +328,20 @@ def isSulfur(self):
+
+    def isIodine(self):
+        """
+        Return ``True`` if the atom represents an sulfur atom or ``False`` if
```
Same here, but for "Iodine"

amarkpayne commented on rmgpy/molecule/group.py:
```diff
@@ -1117,18 +1117,25 @@ def updateFingerprint(self):
         carbon   = atomTypes['C']
         nitrogen = atomTypes['N']
         oxygen   = atomTypes['O']
+        # flourine = atomTypes['F']
```
I think we should either go ahead and add fluorine and iodine atom types in this PR, or remove this commented line until we are ready.

alongd commented:
Thanks for the timely review, @amarkpayne! I addressed all comments but one in fixup commits. I'll double check the implications of removing the `if sum_is_atom == 1:` condition.

alongd commented:
I just saw your comment. I pushed an optional solution, if it won't do the trick I'll squash w/o it

alongd commented:
We believe this PR is a comprehensive example for adding new reactive elements into RMG.
See also https://github.com/ReactionMechanismGenerator/RMG-Py/pull/1206 for sulfur and https://github.com/ReactionMechanismGenerator/RMG-Py/pull/888 for nitrogen
